inference-engine/src/mkldnn_plugin/nodes/mkldnn_detection_output_node.cpp

```diff
@@ -126,480 +153,755 @@ void MKLDNNDetectionOutputNode::initSupportedPrimitiveDescriptors() {
-                dst_data[count * DETECTION_SIZE + 2] = pconf[c*_num_priors + idx];
-
-                float xmin = _share_location ? pboxes[idx*4 + 0] :
-                             pboxes[c*4*_num_priors + idx*4 + 0];
-                float ymin = _share_location ? pboxes[idx*4 + 1] :
-                             pboxes[c*4*_num_priors + idx*4 + 1];
-                float xmax = _share_location ? pboxes[idx*4 + 2] :
-                             pboxes[c*4*_num_priors + idx*4 + 2];
-                float ymax = _share_location ? pboxes[idx*4 + 3] :
-                             pboxes[c*4*_num_priors + idx*4 + 3];
-
-                if (_clip_after_nms) {
-                    xmin = (std::max)(0.0f, (std::min)(1.0f, xmin));
-                    ymin = (std::max)(0.0f, (std::min)(1.0f, ymin));
-                    xmax = (std::max)(0.0f, (std::min)(1.0f, xmax));
-                    ymax = (std::max)(0.0f, (std::min)(1.0f, ymax));
+inline void MKLDNNDetectionOutputNode::confReorderDense(const float *confData, const float *ARMConfData, float *reorderedConfData) {
+    if (withAddBoxPred) {
+        parallel_for2d(imgNum, priorsNum, [&](size_t n, size_t p) {
+            if (ARMConfData[n * priorsNum * 2 + p * 2 + 1] < objScore) {
+                for (int c = 0; c < classesNum; ++c) {
+                    reorderedConfData[n * priorsNum * classesNum + c * priorsNum + p] = c == backgroundClassId ? 1.0f : 0.0f;
+                }
+            } else {
+                for (int c = 0; c < classesNum; ++c) {
+                    reorderedConfData[n * priorsNum * classesNum + c * priorsNum + p] = confData[n * priorsNum * classesNum + p * classesNum + c];
                 }
-
-                dst_data[count * DETECTION_SIZE + 3] = xmin;
-                dst_data[count * DETECTION_SIZE + 4] = ymin;
-                dst_data[count * DETECTION_SIZE + 5] = xmax;
-                dst_data[count * DETECTION_SIZE + 6] = ymax;
-
-                ++count;
             }
-        }
+        });
+        return;
     }
-
-    if (count < num_results) {
-        // marker at end of boxes list
-        dst_data[count * DETECTION_SIZE + 0] = -1;
+    // withAddBoxPred is false
+    if (permuteKernel_) {
+        auto srcData = reinterpret_cast<const uint8_t*>(confData);
+        auto dstData = reinterpret_cast<uint8_t*>(reorderedConfData);
+        permuteKernel_->execute(srcData, dstData);
+        return;
     }
-}
-
-struct ConfidenceComparator {
-    explicit ConfidenceComparator(const float* conf_data) : _conf_data(conf_data) {}
-
-    bool operator()(int idx1, int idx2) {
-        if (_conf_data[idx1] > _conf_data[idx2]) return true;
-        if (_conf_data[idx1] < _conf_data[idx2]) return false;
-        return idx1 < idx2;
+    for (int n = 0; n < imgNum; ++n) {
+        for (int c = 0; c < classesNum; ++c) {
+            for (int p = 0; p < priorsNum; ++p) {
```

Comment:
I think we could use `parallel_for3d` for these cycles

Reply:
done, thank you.